dev/Notifications/BadgeNotifications/BadgeNotificationManager.cpp

```diff
@@ -0,0 +1,157 @@
+        {
+            return storedBadgeNotificationManager.as<winrt::Microsoft::Windows::BadgeNotifications::BadgeNotificationManager>();
+        }
+        else
+        {
+            // Need to clear the RoActivateInstance caching for the PushNotificationLongRunningProcess proxyStub to be found.
+            ::Microsoft::RoApi::ClearRoActivateInstanceCache();
+
+            // Store the BadgeNotificationManager in the COM static store
+            auto badgeNotificationManager{ winrt::make<BadgeNotificationManager>() };
+            appProperties.Insert(STORED_BADGENOTIFICATION_MANAGER_KEY, badgeNotificationManager);
+            return badgeNotificationManager;
+        }
+    }
+
+    void BadgeNotificationManager::SetBadgeAsCount(uint32_t notificationCount)
+    {
+        THROW_HR_IF_MSG(E_ILLEGAL_METHOD_CALL, !AppModel::Identity::IsPackagedProcess(), "Not applicable for unpackaged applications");
+
+        SetBadge(std::to_wstring(notificationCount), nullptr);
+        return;
+    }
+
+    void BadgeNotificationManager::SetBadgeAsCount(uint32_t notificationCount, winrt::DateTime expiration)
+    {
+        THROW_HR_IF_MSG(E_ILLEGAL_METHOD_CALL, !AppModel::Identity::IsPackagedProcess(), "Not applicable for unpackaged applications");
+
+        SetBadge(std::to_wstring(notificationCount), &expiration);
+        return;
+    }
+
+    void BadgeNotificationManager::SetBadgeAsGlyph(winrt::BadgeNotificationGlyph glyphValue)
+    {
+        THROW_HR_IF_MSG(E_ILLEGAL_METHOD_CALL, !AppModel::Identity::IsPackagedProcess(), "Not applicable for unpackaged applications");
+
+        PCWSTR glyphValueString;
+        GetBadgeNotificationGlyphToString(glyphValue, &glyphValueString);
+        SetBadge(glyphValueString, nullptr);
+        return;
+    }
+
+    void BadgeNotificationManager::SetBadgeAsGlyph(winrt::BadgeNotificationGlyph glyphValue, winrt::DateTime expiration)
+    {
+        THROW_HR_IF_MSG(E_ILLEGAL_METHOD_CALL, !AppModel::Identity::IsPackagedProcess(), "Not applicable for unpackaged applications");
+
+        PCWSTR glyphValueString;
+        GetBadgeNotificationGlyphToString(glyphValue, &glyphValueString);
+        SetBadge(glyphValueString, &expiration);
+        return;
+    }
+
+    void BadgeNotificationManager::ClearBadge()
+    {
+        THROW_HR_IF_MSG(E_ILLEGAL_METHOD_CALL, !AppModel::Identity::IsPackagedProcess(), "Not applicable for unpackaged applications");
+
+        auto logTelemetry{ BadgeNotificationTelemetry::ClearBadge::Start(
+            g_telemetryHelper,
+            m_appId) };
+
+        BaseNotifications::BaseNotificationManager::RemoveAllNotification(ToastABI::NotificationType::NotificationType_Badge);
+
+        logTelemetry.Stop();
+
+        return;
+    }
+
+    void BadgeNotificationManager::GetBadgeNotificationGlyphToString(_In_ winrt::BadgeNotificationGlyph glyphValue, _Out_ PCWSTR* glyphString)
+    {
+        static const std::unordered_map<winrt::BadgeNotificationGlyph, PCWSTR> enumMapping = {
```

Comment:
Don't build a static map like this - use an array, with lookup.

For example:

```c++
constexpr static std::pair<winrt::BadgeNotificationGlyph, PCWSTR> mapping[] = {
    { winrt::BadgeNotificationGlyph::None, L"none" },
    { winrt::BadgeNotificationGlyph::Activity, L"activity" },
    // ...
};
auto i = std::find_if(std::begin(mapping), std::end(mapping), [&](auto const& i) { return i.first == glyphValue; });
if (i != std::end(mapping)) {
    *glyphString = i->second;
}
```

That's compile-time initialized.  The linear scan of an array, amortized, along with the reduction in binary size, vastly outweighs the "O(1) lookup" promise.

Or use a `switch(...)` statement.  Just don't use a `static std::[unordered_]map`.

Reply:
changed to array